hibernate-core/src/main/java/org/hibernate/query/sqm/tree/predicate/SqmLikePredicate.java

```diff
@@ -57,6 +59,9 @@ public SqmLikePredicate(
 		);
 		matchExpression.applyInferableType( expressibleType );
 		pattern.applyInferableType( expressibleType );
+
+		assertComparable( matchExpression, pattern, nodeBuilder.getSessionFactory() );
```

Comment:
I don't know how others feel about the type check here, but I think I'd rather like to see something like

```suggestion
		assertString( matchExpression );
		assertString( pattern );
```

I don't know how strict the `assertString` should be, maybe something like the following:

```java
	public static void assertString(SqmExpression<?> expression) {
		final SqmExpressible<?> nodeType = expression.getNodeType();
		if ( nodeType != null ) {
			final Class<?> javaType = nodeType.getExpressibleJavaType().getJavaTypeClass();
			if ( javaType != String.class && javaType != char[].class ) {
				throw new SemanticException( "Operand of 'like' is of type '" + nodeType.getTypeName() + "' which is not a string"
						+ " (it is not an instance of 'java.lang.String' or 'char[]')" );
			}
		}
	}
```

Reply:
I'd say that it's the same amount of strict as assertNumeric or assertDuration so shouldn't be an issue